src/components/Canvas/interactionLayer.tsx

```diff
@@ -0,0 +1,505 @@
+import { PixelChangeRecords } from "../../helpers/PixelChangeRecords";
+import {
+  addColumnToData,
+  addRowToData,
+  deleteColumnOfData,
+  deleteRowOfData,
+  extractColoredPixelsFromColumn,
+  extractColoredPixelsFromRow,
+  getColumnCountFromData,
+  getGridIndicesFromData,
+  getRowCountFromData,
+} from "../../utils/data";
+import { convertCartesianToScreen, getScreenPoint } from "../../utils/math";
+import { BaseLayer } from "./BaseLayer";
+import {
+  DefaultGridSquareLength,
+  UserId,
+  ButtonDirection,
+  TemporaryUserId,
+} from "./config";
+import {
+  ColorChangeItem,
+  Coord,
+  DottingData,
+  GridIndices,
+  PixelModifyItem,
+} from "./types";
+
+export default class InteractionLayer extends BaseLayer {
+  // We make this a map to allow for multiple users to interact with the canvas
+  // the key will be the user id
+  private strokedPixelRecords: Map<UserId, PixelChangeRecords> = new Map();
+
+  private erasedPixelRecords: Map<UserId, PixelChangeRecords> = new Map();
+
+  // this is for tracking pixels that were applied to data canvas
+  // while the user was changing the dimensions of the data canvas
+  // ( = the captured Data is not null)
+  private tempStrokedPixels: Array<PixelModifyItem> = [];
+
+  private swipedPixels: Array<PixelModifyItem> = [];
```

Comment:
Having duplicated swipedPixels in dataLayer and InteractionLayer confuses me

Reply:
Yes, yes. This design is very confusing at first. Especially the interaction layer is the tricky part. However as you have stated, the swiped pixels do not necessarily need to be in interaction layer. I will erase it!